{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "453a0f47",
   "metadata": {},
   "source": [
    "# Extract UbiNet presumed degrons "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af7444b",
   "metadata": {},
   "source": [
    "This notebook contains the code to extract each UbiNet E3-ligase presumed degrons, meaning those protein subsequences used in UbiNet for each motif construction (MSA). Specifically:\n",
    "1. First, the MSAs from the HTML of each motif are parsed to extract the degrons. As UbiNet IDs are build using the structure `E3-AC6_SUBS-AC5` (AC5 meaning the accession number contains only 5 chracters), an additional matching step with the ESIs dictionary and the sequence in the proteome is required to correctly assign the AC to the substrate.\n",
    "2. Second, MSA visualization to identify those alignments with indels.\n",
    "3. Third, remove substrate sequences whose length differ from the motif's length (we consider these substrates are non-matrix-contributing sequences)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd2dd7ac",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2befb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "from bs4 import BeautifulSoup\n",
    "import requests  \n",
    "import gzip\n",
    "import json\n",
    "\n",
    "## my modules ##\n",
    "sys.path.append(\"../Utils/\")    # modules folder\n",
    "from fasta_utils import readFasta_gzip\n",
    "from motifs_utils import motif_scan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fe9c11d",
   "metadata": {},
   "source": [
    "## Define variables and paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e426a6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "base = \"../../\"\n",
    "\n",
    "data = \"data/\"\n",
    "\n",
    "url_motifs_path = os.path.join(base, data, \"ubinet/motif_links_ubinet.txt\")\n",
    "ubinet_degrons_path = os.path.join(base, data, \"ubinet/all_positive_set/\")\n",
    "ubinet_degrons_kept_path = os.path.join(base, data, \"ubinet/kept_positive_set/\")\n",
    "ubinet_degrons_discarded_path = os.path.join(base, data, \"ubinet/discarded_positive_set/\")\n",
    "weight_m_path = os.path.join(base, data, \"ubinet/motif_matrices/PWM/\")                 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b889bdbb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "41d85b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test OK!\n",
    "def map_E3_substrates(ESIs_file, scan_dir, ESIs_file_filtered = None, sep = \"\\t\", save = True):\n",
    "    \"\"\"\n",
    "    Generates a human-filtered E3-ligase-substrate interactions dictionary \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    ESIs_file: str\n",
    "                Path to the folder where the file containing E3-ligase-substrate interactions (ESIs) is located \n",
    "    scan_dir: str\n",
    "                Path to the folder where the PWM scans are stored\n",
    "    ESIs_file_filtered: str (default: None)\n",
    "                Path to the folder where the filtered ESIs file will be located. Not mandatory. \n",
    "    sep: str (default: \"\\t\")\n",
    "                Column separator in the ESIs_file\n",
    "    save: boolean (default: True)\n",
    "                If True, the ESIs dictionary is considered to be saved. \n",
    "                If False, the ESIs dictionary is not considered to be saved. \n",
    "                By default, the function considers saving the ESIs dictionary as a file, \n",
    "                printing an error bypass message in case no ESIs_file_filtered is provided. \n",
    "                \n",
    "    Returns\n",
    "    -------\n",
    "    ESIs: dict\n",
    "                Dictionary containing E3-ligase-substrate interactions (ESIs) in the form of \n",
    "                {E3-ligase AC: [substrates AC]}\n",
    "    \"\"\"\n",
    "    \n",
    "    # Load E3-substrate interactions file\n",
    "    ESIs_df = pd.read_csv(ESIs_file, sep = sep)\n",
    "    \n",
    "    # Generate a dictionary of E3-ligases and their corresponding substrates\n",
    "    E3s = [E3.split(\".\")[0] for E3 in os.listdir(scan_dir)]\n",
    "    ESIs = {}\n",
    "    counter = 0\n",
    "    for E3 in E3s:\n",
    "        #print(f'E3-ligase: {E3}')\n",
    "        substrates = ESIs_df.loc[(ESIs_df.E3_AC == E3) & \n",
    "                                 (ESIs_df.SUB_ORGANISM == 'Homo sapiens (Human)') & (ESIs_df.SUB_AC != '-'),\n",
    "                                 \"SUB_AC\"].values\n",
    "        substrates = list(set(substrates))     # set transformation to avoid duplicated substrates \n",
    "                                               # (problem with UbiNet data)\n",
    "        ESIs[E3] = substrates\n",
    "        counter += 1\n",
    "    \n",
    "    if save:\n",
    "        try:\n",
    "            with open(ESIs_file_filtered+'.json', 'w') as fp:\n",
    "                json.dump(ESIs, fp)\n",
    "        except TypeError:\n",
    "            print(\"Error: name for output file not provided, saving aborted. \\\n",
    "            The dictionary is still generated as a variable.\")\n",
    "    \n",
    "    print(f'Number of retrieved ESIs: {counter}')\n",
    "    return ESIs\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2303bcc9-c653-4278-baab-9eb6a4a54e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test OK!\n",
    "def parse_motifs_msa(urlbase, links_path, ESIs, proteome, pos_set_dir):\n",
    "    \"\"\"\n",
    "    Parses a UbiNet motif's MSA from the E3 ligase motif HTML and extracts a MSA dataframe per E3-ligase \n",
    "    containing gene, sequence, start and end columns. Replaces gene's 5-characters ACs with its corresponding \n",
    "    6-character AC\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    urlbase: str\n",
    "                Non-mutable part of the URL to access the website\n",
    "    links_path: str\n",
    "                Path to the folder where the text file with all motifs hyperlinks will be located\n",
    "    ESIs: dict\n",
    "                Dictionary containing E3-ligase-substrate interactions (ESIs) in the form of {E3-ligase AC: \n",
    "                [substrates ACs]}\n",
    "    proteome: dict\n",
    "                Dictionary containing each protein's accession number as key and its sequence as value\n",
    "    pos_set_dir: str\n",
    "                Path to the folder where the parsed MSAs will be stored, named after the corresponding E3-ligase AC\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    \n",
    "    counter = 0                       # (optional) Monitor progress\n",
    "    \n",
    "\n",
    "    with open(links_path) as file:\n",
    "        for url in file:\n",
    "            \n",
    "            counter += 1              # (optional) Monitor progress\n",
    "            \n",
    "            E3 = url.split(\"/\")[3]\n",
    "            print(f'{counter}: {E3}')      # (optional) Monitor progress\n",
    "            \n",
    "            \n",
    "            # Extract motifs MSA\n",
    "            r = (requests.get(urlbase+url.strip())).text              \n",
    "            soup = BeautifulSoup(r, 'html.parser')\n",
    "            msa = str(soup.find_all(name = 'input', attrs = {\"name\": \"aln1\"}))  # string transformation required \n",
    "                                                                                # for further processing\n",
    "            \n",
    "            \n",
    "            # MSA table processing\n",
    "            msa = msa.split(\"*\")[-1].split(\"\\n\")[1:-1]                          # splitting by *, retrieve matrix \n",
    "                                                                                # only; then, splitting by \\n, \n",
    "                                                                                # separate matrix rows; indeces to \n",
    "                                                                                # keep rows only\n",
    "            msa = [row.split() for row in msa]                                  # separate matrix cols\n",
    "            msa_df = pd.DataFrame(msa).iloc[:, :4]                              # create df with cols: gene, \n",
    "                                                                                # sequence, start, end (first four)\n",
    "            msa_df.rename(columns = {0: \"gene\", \n",
    "                                     1: \"start\", \n",
    "                                     2: \"sequence\", \n",
    "                                     3: \"end\"}, \n",
    "                          inplace = True)\n",
    "            msa_df = msa_df[[\"gene\", \"sequence\", \"start\", \"end\"]]               # reorder cols\n",
    "            msa_df[\"gene\"] = [gene.split(\"_\")[1] \n",
    "                              for gene in msa_df[\"gene\"]]                       # keep substrate 5-characters AC \n",
    "                                                                                # (remove E3-ligase AC from the \n",
    "                                                                                # name)\n",
    "            msa_df[\"sequence\"] = [seq.replace('.', '') \n",
    "                                  for seq in msa_df[\"sequence\"]]                # remove MSA indels to obtain \n",
    "                                                                                # actual sequence\n",
    "            \n",
    "            \n",
    "            # MSA table processing: complete gene ACs (using flag variable)\n",
    "            # Check E3s independently \n",
    "            ESIs_E3 = ESIs[E3]                                                  # E3-ligase substrates ACs \n",
    "                                                                                # (according to UbiNet ESIs \n",
    "                                                                                # downloadable table)\n",
    "            \n",
    "            for msa_gene in msa_df[\"gene\"]:                                     # Filter those 5-characters ACs \n",
    "                                                                                # with more than 1 match in the \n",
    "                                                                                # E3-ligase ESIs list \n",
    "                                                                                # (ambigous genes)\n",
    "                # Check every E3 substrates in the ESIs dictionary\n",
    "                flag = 0\n",
    "                for substrate in ESIs_E3:\n",
    "                    if msa_gene == substrate[:len(msa_gene)]:\n",
    "                        flag += 1\n",
    "                \n",
    "                # When a 5-characters AC has a single match, replace it with the matching 6-characters AC\n",
    "                if flag == 1:                                                       \n",
    "                    for substrate in ESIs_E3:\n",
    "                        if msa_gene == substrate[:len(msa_gene)]:\n",
    "                            msa_df.replace({msa_gene: substrate}, inplace = True)\n",
    "\n",
    "                # When a 5-characters AC has more than one match, identify the sequence in the proteome and \n",
    "                # replace it with the matching 6-characters AC (here, we are assuming one match is always \n",
    "                # going to happen)\n",
    "                else: \n",
    "                    \n",
    "                    matching_substrates = []\n",
    "                    \n",
    "                    # Find all posible matching substrates in ESIs dictionary\n",
    "                    for substrate in ESIs_E3:\n",
    "                        if msa_gene == substrate[:len(msa_gene)]:\n",
    "                            matching_substrates.append(substrate)\n",
    "                    #print(f'Substrates in the ESIs dictionary matching the ambigous df gene: {matching_substrates}')\n",
    "                    \n",
    "                    # Find all possible matching substrates in the MSA dataframe (sometimes, there is ambiguity\n",
    "                    # also here)\n",
    "                    ambigous_msa_df = msa_df.loc[msa_df[\"gene\"] == msa_gene]\n",
    "                    #print(f'\\nDataframe subset containing the ambigous genes rows: {ambigous_msa_df}')\n",
    "                    \n",
    "                    # Retrieve ambigous substrates sequences\n",
    "                    for index, row in ambigous_msa_df.iterrows():\n",
    "                        seq, start, end = (row[\"sequence\"], row[\"start\"], row[\"end\"])\n",
    "                        \n",
    "                        # Retrieve sequences from all possible matching substrates\n",
    "                        for matching_substrate in matching_substrates:\n",
    "                            #print(f'\\nMatching dictionary substrate being analyzed: {matching_substrate}')\n",
    "\n",
    "                            seq_proteome = proteome[matching_substrate][int(start)-1:int(end)]\n",
    "                            #print(f'\\nProteome sequence: {seq_proteome}')\n",
    "                            #print(f'Dataframe sequence: {seq}')\n",
    "                            \n",
    "                            # Check whether the MSA sequence and the proteome-extracted sequence are the same\n",
    "                            if seq == seq_proteome:\n",
    "                                #print(f\"{matching_substrate} Match!! Updating dataframe\")\n",
    "                                msa_df.update(msa_df.loc[(msa_df[\"start\"] == start) & (msa_df[\"end\"] == end)].\n",
    "                                              replace({msa_gene: matching_substrate}))\n",
    "                            #elif seq != seq_proteome:\n",
    "                                #print(f\"{matching_substrate} No match :(\")\n",
    "                            \n",
    "            # Store MSA sequences (tsv file per E3-ligase)    \n",
    "            #print(f'Formated MSA sequences dataframe\\n {msa_df}')\n",
    "            msa_df.to_csv(pos_set_dir+E3+\".tsv\", sep = \"\\t\", header = True, index = False)\n",
    "            \n",
    "    return None\n",
    "\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "804f124c",
   "metadata": {},
   "source": [
    "Note on the rule to find the MSA on the HTML soup: below, the HTML part where the MSA is located.\n",
    "\n",
    "`<input type=\"hidden\" name=\"aln1\" value=\".\n",
    "                  ********\n",
    "Q86YT6_Q9UHD  469 DFCIRNIE  476 + 13.3\n",
    "Q86YT6_P5413 1053 DFCKKHPD 1060 + 14.5\n",
    "Q86YT6_P7850  482 DHCERDID  489 + 19.9\n",
    "Q86YT6_O0054  400 FNCEKKID  407 + 13.9\n",
    "Q86YT6_P5335 1089 DICARDLS 1096 + 9.77\n",
    "Q86YT6_Q86YT  993 PICRKAIE 1000 + 14.0\n",
    "Q86YT6_Q9NYJ  348 SNCEKRVD  355 + 14.1\n",
    "Q86YT6_Q9NR6  397 SNCEKKVD  404 + 16.0\n",
    "Q86YT6_Q9UPN  746 QRCLRQAE  753 + 8.71\n",
    "Q86YT6_O1507 2053 DHCQREQE 2060 + 12.0\n",
    "Q86YT6_Q1515 1702 QRCKRKIE 1709 + 17.2\n",
    "Q86YT6_Q9UNE   85 QICRRHKD   92 + 15.0\n",
    "Q86YT6_Q1663  248 PICPDSLD  255 + 3.88\n",
    ".\">`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9cc7317",
   "metadata": {},
   "source": [
    "According to Beautiful Soup documentation, `name` corresponds to the tag at the beggining of the 'less than' sign. Attributes are whatever comes next, and MSA tables always contein `{\"name\": \"aln1\"}`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "44910ee0-4fcb-4510-b114-bb2018334a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_motifs_msa(pos_set_dir, weight_m_dir, pos_set_filtered_dir, pos_set_discarded_dir):\n",
    "    \"\"\"\n",
    "    Filters a dataframe of the type (gene, sequence, start, end) to only keep those sequences of the same\n",
    "    length as the PWM. If applicable, generates two files of kept and discarded sequences\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    pos_set_dir: str\n",
    "                Path to the folder where a dataframe of the type (gene, sequence, start, end) (positive set)\n",
    "                is located\n",
    "    weight_m_dir: str\n",
    "                Path to the folder where the weight matrices are located. Each file has to be tab-separated\n",
    "    pos_set_filtered_dir: str\n",
    "                Path to the folder where the kept dataframe of the type (gene, sequence, start, end) (positive set)\n",
    "                is located\n",
    "    pos_set_discarded_dir: str\n",
    "                Path to the folder where the discarded dataframe of the type (gene, sequence, start, end) \n",
    "                (positive set) is located\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    \n",
    "    #counter = 0                       # (optional) Monitor progress\n",
    "    \n",
    "    E3s = os.listdir(pos_set_dir)\n",
    "    \n",
    "    for E3 in E3s:\n",
    "        \n",
    "        #counter += 1                   # (optional) Monitor progress\n",
    "        #print(f'{counter}: {E3}')      # (optional) Monitor progress\n",
    "        \n",
    "        # Load MSA positive sequences and PWM of the specific E3 ligase\n",
    "        msa_df = pd.read_csv(pos_set_dir+E3, sep = \"\\t\")\n",
    "        weight_m = pd.read_csv(weight_m_dir+E3, sep = \"\\t\")\n",
    "        \n",
    "        # Filter the positive sequences to generate kept and discarded sequences according to PWM length\n",
    "        msa_filtered_df = msa_df.loc[msa_df.sequence.apply(lambda x: len(x) == len(weight_m))]\n",
    "        msa_discarded_df = msa_df.loc[msa_df.sequence.apply(lambda x: len(x) != len(weight_m))]\n",
    "    \n",
    "        msa_filtered_df.to_csv(pos_set_filtered_dir+E3, sep = \"\\t\", header = True, index = False)\n",
    "        if not msa_discarded_df.empty:\n",
    "            msa_discarded_df.to_csv(pos_set_discarded_dir+E3, sep = \"\\t\", header = True, index = False)\n",
    "    \n",
    "    return None\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38d04b72",
   "metadata": {},
   "source": [
    "## 1. Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "096a0f73-3d61-4020-9039-9f643e44c76f",
   "metadata": {},
   "source": [
    "### 1.1. Generate ESIs dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f0344fa8-d6c2-4f46-bad2-0b4f375811fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of retrieved ESIs: 104\n"
     ]
    }
   ],
   "source": [
    "ESIs = map_E3_substrates(data_path+external_data_path+ESIs_file, data_path+scan_path, \n",
    "                         data_path+ESIs_file_filtered, save = False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab17cac-13f8-4f16-a5bd-7594c9c9e2ac",
   "metadata": {},
   "source": [
    "### 1.2. Proteome loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9aa29965-dd33-4510-b5f3-7a7e077825e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of retrieved sequences: 78120\n"
     ]
    }
   ],
   "source": [
    "proteome = readFasta_gzip(data_path+external_data_path+proteome_file)\n",
    "\n",
    "# Expected number of read sequences: 78120 (https://www.uniprot.org/proteomes/UP000005640)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790389de-0d28-4185-bae3-4c7b1d40aa42",
   "metadata": {},
   "source": [
    "## 2. MSAs parsing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7ccd4f",
   "metadata": {},
   "source": [
    "First, parse the MSAs and generate the presumed degrons dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d00da5a5-0e16-41ff-aa2e-22de2b0ae3b8",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1: Q96AX9\n",
      "2: Q13191\n",
      "3: Q8WZ73\n",
      "4: Q8TBB1\n",
      "5: O76064\n",
      "6: Q969Q1\n",
      "7: Q13490\n",
      "8: Q7Z6J0\n",
      "9: Q9BYM8\n",
      "10: Q8WVD3\n",
      "11: Q13049\n",
      "12: Q9Y4K3\n",
      "13: Q13233\n",
      "14: Q9Y4L5\n",
      "15: Q969K3\n",
      "16: Q6ZNA4\n",
      "17: Q8TCQ1\n",
      "18: Q86YT6\n",
      "19: Q5T0T0\n",
      "20: O43255\n",
      "21: Q92844\n",
      "22: Q96EP1\n",
      "23: Q14258\n",
      "24: Q9UKV5\n",
      "25: Q9UHC7\n",
      "26: Q86TM6\n",
      "27: Q9NX47\n",
      "28: Q96EQ8\n",
      "29: Q00987\n",
      "30: Q96F44\n",
      "31: Q9NVW2\n",
      "32: Q13489\n",
      "33: Q96PM5\n",
      "34: P22681\n",
      "35: Q8TEB7\n",
      "36: Q99942\n",
      "37: Q8IUQ4\n",
      "38: Q86YJ5\n",
      "39: P78317\n",
      "40: Q6VVB1\n",
      "41: Q969V5\n",
      "42: P98170\n",
      "43: P29372\n",
      "44: Q8NG27\n",
      "45: Q9H4P4\n",
      "46: Q96FA3\n",
      "47: Q9BV68\n",
      "48: O60858\n",
      "49: Q9NTX7\n",
      "50: P35226\n",
      "51: Q99728\n",
      "52: O43164\n",
      "53: Q96PU5\n",
      "54: O95071\n",
      "55: Q96J02\n",
      "56: Q9H0M0\n",
      "57: Q8IYU2\n",
      "58: Q9HCE7\n",
      "59: Q14669\n",
      "60: P46934\n",
      "61: Q5T447\n",
      "62: Q9HAU4\n",
      "63: Q7Z6Z7\n",
      "64: O00308\n",
      "65: Q05086\n",
      "66: Q9UNE7\n",
      "67: O95155\n",
      "68: O60260\n",
      "69: Q9NWF9\n",
      "70: Q96EP0\n",
      "71: Q12834\n",
      "72: Q9UM11\n",
      "73: O43791\n",
      "74: Q9H2C0\n",
      "75: Q14145\n",
      "76: Q9Y2M5\n",
      "77: Q53G59\n",
      "78: Q7L5Y6\n",
      "79: Q9NZJ0\n",
      "80: Q9Y4B6\n",
      "81: Q96SW2\n",
      "82: Q969H0\n",
      "83: Q13309\n",
      "84: Q9UK22\n",
      "85: Q969P5\n",
      "86: Q9Y297\n",
      "87: Q9UKB1\n",
      "88: Q9UKA1\n",
      "89: P0C2W1\n",
      "90: Q5XUX0\n",
      "91: Q9UKT5\n",
      "92: Q9NXK8\n",
      "93: Q9UKC9\n",
      "94: P19474\n",
      "95: O14543\n",
      "96: O15524\n",
      "97: P40337\n",
      "98: Q12933\n",
      "99: Q99496\n",
      "100: P38398\n",
      "101: Q9NS91\n",
      "102: Q8N3Y1\n",
      "103: P49427\n",
      "104: P10523\n"
     ]
    }
   ],
   "source": [
    "urlbase = \"https://awi.cuhk.edu.cn/~ubinet\"\n",
    "\n",
    "parse_motifs_msa(urlbase, url_motifs_path, ESIs, proteome, ubinet_degrons_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69645e28-be2d-4dae-8e42-c8068f39782f",
   "metadata": {},
   "source": [
    "#### Motif's MSA sequences length range"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a33fd56",
   "metadata": {},
   "source": [
    "After visualizing some E3 ligases MSAs, we realized sometimes they were created from sequences of different length. To avoid this, we thought of considering sequences of different length of the motif to be non-contributing to the PWM. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f52b0306",
   "metadata": {},
   "source": [
    "First, we checked the number of E3 ligases motifs built from sequences of different length:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bb7404aa-dfd2-41ac-b246-802c4a216bb5",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q96AX9.tsv motif's MSA sequences length range is (8, 9)\n",
      "Q7Z6J0.tsv motif's MSA sequences length range is (9, 11)\n",
      "Q9BYM8.tsv motif's MSA sequences length range is (9, 10)\n",
      "Q9Y4L5.tsv motif's MSA sequences length range is (9, 10)\n",
      "O43255.tsv motif's MSA sequences length range is (7, 16)\n",
      "Q92844.tsv motif's MSA sequences length range is (7, 8)\n",
      "Q9UHC7.tsv motif's MSA sequences length range is (7, 8)\n",
      "Q9NVW2.tsv motif's MSA sequences length range is (7, 8)\n",
      "P22681.tsv motif's MSA sequences length range is (9, 10)\n",
      "Q99942.tsv motif's MSA sequences length range is (8, 9)\n",
      "Q8IUQ4.tsv motif's MSA sequences length range is (6, 7)\n",
      "Q969V5.tsv motif's MSA sequences length range is (9, 11)\n",
      "Q9NTX7.tsv motif's MSA sequences length range is (9, 10)\n",
      "Q96PU5.tsv motif's MSA sequences length range is (7, 12)\n",
      "Q96J02.tsv motif's MSA sequences length range is (7, 8)\n",
      "O00308.tsv motif's MSA sequences length range is (8, 33)\n",
      "O43791.tsv motif's MSA sequences length range is (10, 29)\n",
      "Q53G59.tsv motif's MSA sequences length range is (10, 20)\n",
      "Q7L5Y6.tsv motif's MSA sequences length range is (9, 10)\n",
      "Q9Y4B6.tsv motif's MSA sequences length range is (9, 10)\n",
      "Q96SW2.tsv motif's MSA sequences length range is (6, 7)\n",
      "Q969P5.tsv motif's MSA sequences length range is (9, 10)\n",
      "P19474.tsv motif's MSA sequences length range is (8, 9)\n",
      "\n",
      "23 out of 104 E3-ligases motifs were created from very different length degrons sequences\n"
     ]
    }
   ],
   "source": [
    "E3s = os.listdir(ubinet_degrons_path)\n",
    "\n",
    "counter = 0                        # monitor\n",
    "\n",
    "# Dictionary to store those E3 ligases whose positive sequences have different length\n",
    "lowq_E3s = {}\n",
    "\n",
    "for E3 in E3s:\n",
    "    \n",
    "    # Load MSA df\n",
    "    msa_df = pd.read_csv(ubinet_degrons_path+E3, sep = \"\\t\")\n",
    "    \n",
    "    # Check longest and shortest sequences\n",
    "    max_len = msa_df.sequence.str.len().max()\n",
    "    min_len = msa_df.sequence.str.len().min()\n",
    "        \n",
    "    if min_len != max_len:\n",
    "        lowq_E3s[E3.split(\".\")[0]] = (min_len, max_len)\n",
    "        print(f\"{E3} motif's MSA sequences length range is {(min_len, max_len)}\")\n",
    "        counter += 1\n",
    "\n",
    "print(f\"\\n{counter} out of {len(E3s)} E3-ligases motifs were created from very different length degrons sequences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb179e2-cbed-443d-bc2b-cd37970576e1",
   "metadata": {},
   "source": [
    "#### Alignment visualization for MSAs with different length sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41d27907",
   "metadata": {},
   "source": [
    "Next, we visualize the alignments to understand what happens. Here is when we understood that the motif's length was the length of the majority of sequences, and used this rule to discard the non-contributing sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "747e2c55-4216-4583-9c43-caff99cd0994",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1: Q96AX9\n",
      "\tMotif's length: 9\n",
      "\n",
      "                  *********\n",
      "Q96AX9_Q9NQC  105 NCEERFSLF  113 + 16.2\n",
      "Q96AX9_Q9NR6  398 NCEKKVDRC  406 + 26.8\n",
      "Q96AX9_Q9NYJ  349 NCEKRVDRC  357 + 26.4\n",
      "Q96AX9_O0054  401 NCEKKIDYC  409 + 23.0\n",
      "Q96AX9_P4798  992 NCWKKRGLC 1000 + 16.3\n",
      "Q96AX9_Q9Y6K   69 N.QILRERC   76 + 3.74\n",
      "Q96AX9_Q9UHD  578 N.EEQIHKF  585 + 7.43\n",
      "\n",
      "\tMSA sequences length range: (8, 9)\n",
      "\n",
      "2: Q7Z6J0\n",
      "\tMotif's length: 10\n",
      "\n",
      "                 ****.******\n",
      "Q7Z6J0_Q1501 279 SILY.FYSSLS 288 + 9.34\n",
      "Q7Z6J0_Q8WUM 115 SLGY.EKSCVL 124 + 11.2\n",
      "Q7Z6J0_P4804 141 TIGY.GFRCVT 150 + 26.9\n",
      "Q7Z6J0_O1496 693 TMGYMGSQSVS 703 + 7.30\n",
      "Q7Z6J0_P6325 142 TIGY.GFRCVT 151 + 26.9\n",
      "Q7Z6J0_Q9UBP 123 SIG..FRECLT 131 + 10.5\n",
      "\n",
      "\tMSA sequences length range: (9, 11)\n",
      "\n",
      "3: Q9BYM8\n",
      "\tMotif's length: 10\n",
      "\n",
      "                 **********\n",
      "Q9BYM8_Q9Y6K 130 RCQQ.QMAED 138 + 13.2\n",
      "Q9BYM8_Q1425  49 LCPQCRAVYQ  58 + 12.1\n",
      "Q9BYM8_P1725 131 KCDTCDMNVH 140 + 15.0\n",
      "Q9BYM8_P4820 119 ACPT.DLTVD 127 + 8.92\n",
      "Q9BYM8_Q9BYM 446 RCPQCQIVVQ 455 + 22.5\n",
      "Q9BYM8_Q8N5C 702 RCEQCEMPRY 711 + 24.5\n",
      "Q9BYM8_Q9NYJ 683 RCEQCEMPRH 692 + 25.1\n",
      "Q9BYM8_Q1465 166 PCPQ.PLRSP 174 + 5.67\n",
      "Q9BYM8_Q1354 232 ICEQ.QLIMC 240 + 10.2\n",
      "Q9BYM8_O1486 180 QTPQCKLRRY 189 + 12.3\n",
      "Q9BYM8_O7546  93 QCQACRLRKC 102 + 14.4\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "4: Q9Y4L5\n",
      "\tMotif's length: 9\n",
      "\n",
      "                  *****.****\n",
      "Q9Y4L5_P0053  834 VHRDL.AARN  842 + 25.6\n",
      "Q9Y4L5_Q9Y4L  109 THTDFWGARP  118 + 4.85\n",
      "Q9Y4L5_P3688  808 VHRDL.AARN  816 + 25.6\n",
      "Q9Y4L5_P0858 1201 VHRDL.AARN 1209 + 25.6\n",
      "Q9Y4L5_P6107  139 VHATN.SQRP  147 + 2.27\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "5: O43255\n",
      "\tMotif's length: 8\n",
      "\n",
      "                  *****........***\n",
      "O43255_P5503  174 PPG.PSLADALI.SSP  187 + 9.45\n",
      "O43255_Q9BSI  370 PPRARKPVCP...PSL  382 + 10.3\n",
      "O43255_Q1362  867 PPY.P........PPP  873 + 15.7\n",
      "O43255_O9601   85 PEE.PT.......PAP   92 + 9.89\n",
      "O43255_O7592  501 QAS.PVSRT....PSL  511 + 11.0\n",
      "O43255_P3784  117 PVD.PDNEAYEM.PSE  130 + 6.98\n",
      "O43255_Q9266   89 PAE.P........PAL   95 + 12.6\n",
      "O43255_Q9GZT   77 PAP.PAAV.....PPP   86 + 16.1\n",
      "O43255_P2959   34 QPS.PS.......PSP   41 + 17.6\n",
      "O43255_Q1663  194 QPPAPAL......PGP  203 + 13.6\n",
      "O43255_P4314 1314 PPSQPEHSSSEEAPSR 1329 + 13.9\n",
      "O43255_O7537  607 PPP.PL.......PPP  614 + 19.3\n",
      "O43255_P0824  236 PPGAPEKQ.....PAP  246 + 17.4\n",
      "O43255_Q9P21 1140 PPN.PL.......PSR 1147 + 14.0\n",
      "O43255_Q9284  347 PSDAPF.......PSL  355 + 12.3\n",
      "O43255_O1497  404 QAT.P........TSP  410 + 9.03\n",
      "O43255_P0463   82 PAA.PA.......PAP   89 + 14.4\n",
      "O43255_O4359   62 VPR.PGLK.....PAP   71 + 9.04\n",
      "O43255_Q9NS8 1139 PQS.PK.......TPP 1146 + 6.96\n",
      "O43255_Q96B9  340 PPP.P........SAP  346 + 14.5\n",
      "O43255_Q9993   35 PPAQRG.......PPP   43 + 14.4\n",
      "O43255_Q9263   46 PPS.PIAL.....PPL   55 + 15.6\n",
      "O43255_Q9Y6H  770 PPNQPSGDPQQ..PSP  783 + 17.8\n",
      "O43255_P4652   91 PPR.PPKGACKV.PAQ  104 + 7.05\n",
      "O43255_Q1666  508 QSS.PEPNS....PSE  518 + 8.05\n",
      "O43255_P3723  110 PAS.PPY......YSE  118 + 4.30\n",
      "O43255_P1027  375 PPP.P........PPP  381 + 19.8\n",
      "O43255_Q9NRM  471 PAPAPA.......PAP  479 + 18.2\n",
      "O43255_Q96KS  399 PVSQP........PTP  406 + 12.3\n",
      "O43255_P3727  554 QPTQP........PPL  561 + 15.2\n",
      "\n",
      "\tMSA sequences length range: (7, 16)\n",
      "\n",
      "6: Q92844\n",
      "\tMotif's length: 8\n",
      "\n",
      "                 ********\n",
      "Q92844_Q8IUC 118 YQEAVRTL 125 + 10.0\n",
      "Q92844_Q1416 198 FGVTVDLW 205 + 12.1\n",
      "Q92844_O9514 148 IRPNVGFW 155 + 5.24\n",
      "Q92844_Q1354 212 YSFAVVLW 219 + 10.7\n",
      "Q92844_Q1311 414 YNG.VLIW 420 + 7.26\n",
      "Q92844_Q1360 129 FEQAVRIL 136 + 11.9\n",
      "Q92844_Q1575 200 YNQTVDIW 207 + 19.7\n",
      "Q92844_Q1348 595 IKGTVRTF 602 + 8.85\n",
      "Q92844_Q9BVC  20 YDHTVRFW  27 + 18.2\n",
      "\n",
      "\tMSA sequences length range: (7, 8)\n",
      "\n",
      "7: Q9UHC7\n",
      "\tMotif's length: 8\n",
      "\n",
      "                  ********\n",
      "Q9UHC7_Q7Z7B  892 SSPGHPGE  899 + 13.6\n",
      "Q9UHC7_P5409  519 GAPGDPMD  526 + 17.3\n",
      "Q9UHC7_P3893  129 GSPGGPGD  136 + 14.5\n",
      "Q9UHC7_P0463   78 AAP.TPAA   84 + 15.0\n",
      "Q9UHC7_O1474  417 AAV.TPAA  423 + 11.9\n",
      "Q9UHC7_Q8N72   35 AAPGAPAA   42 + 18.1\n",
      "Q9UHC7_P1194    4 SAPSYPMA   11 + 12.4\n",
      "Q9UHC7_P4678  130 SAV.PPGA  136 + 9.72\n",
      "\n",
      "\tMSA sequences length range: (7, 8)\n",
      "\n",
      "8: Q9NVW2\n",
      "\tMotif's length: 8\n",
      "\n",
      "                 ********\n",
      "Q9NVW2_P1694  14 RAS.GQAF  20 + 8.63\n",
      "Q9NVW2_Q0098 475 CPVCRQPI 482 + 12.9\n",
      "Q9NVW2_P5427 356 RAT.ESRI 362 + 5.39\n",
      "Q9NVW2_P5366  25 CASCGQRI  32 + 24.1\n",
      "Q9NVW2_Q9276 409 RAS.DKRI 415 + 11.4\n",
      "Q9NVW2_Q86U7   7 CPGCSSKS  14 + 8.59\n",
      "Q9NVW2_Q96MM 219 CAECGKAF 226 + 15.7\n",
      "Q9NVW2_P2579  94 CASCDKRI 101 + 23.2\n",
      "Q9NVW2_Q9UBR  31 CAGCDQHI  38 + 20.3\n",
      "Q9NVW2_P5045  53 CAGCGGKI  60 + 21.4\n",
      "Q9NVW2_Q9NVW 607 CPICRRAV 614 + 9.45\n",
      "Q9NVW2_P6196  23 CAGCGGKI  30 + 21.4\n",
      "\n",
      "\tMSA sequences length range: (7, 8)\n",
      "\n",
      "9: P22681\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  **********\n",
      "P22681_P1127 1232 TDSWSLEVMS 1241 + 13.1\n",
      "P22681_P3522  371 TDPS.QRLVQ  379 + 5.43\n",
      "P22681_P2798  393 SDPLTFSSVV  402 + 9.42\n",
      "P22681_P0858 1264 SDVWSFGVLL 1273 + 28.9\n",
      "P22681_P0051  420 SDVWAFGVLL  429 + 20.7\n",
      "P22681_P4340  553 SDVWSFGVLM  562 + 28.5\n",
      "P22681_P0806 1193 SDVWSFGVVL 1202 + 27.2\n",
      "P22681_P1623  876 SDVWSYGILL  885 + 28.1\n",
      "P22681_P3596 1086 SDVWSFGVLL 1095 + 28.9\n",
      "P22681_P1136  681 SDVWSFGVLL  690 + 28.9\n",
      "P22681_P4018  584 SDTL.YMVRM  592 + 9.82\n",
      "P22681_P0864  777 SDVVSFRLSV  786 + 14.1\n",
      "P22681_P1549  464 NKKWSHMFLL  473 + 5.50\n",
      "P22681_O1496  239 TSPLSQQSQL  248 + 8.31\n",
      "P22681_Q96B9  468 SVVSSTEKLS  477 + 7.34\n",
      "P22681_Q8WU2   54 SVKWHYLCLR   63 + 10.5\n",
      "P22681_Q0491 1268 SDVWSFGVLL 1277 + 28.9\n",
      "P22681_O9487  298 SSPSSPSRAK  307 + 5.48\n",
      "P22681_P0813   74 SDVVSATEPC   83 + 1.89\n",
      "P22681_O4352    6 SDVSSECDRE   15 + 7.15\n",
      "P22681_P5508  380 SRKSSSYSSS  389 + 7.26\n",
      "P22681_P4873  757 YDVSSQVISQ  766 + 8.58\n",
      "P22681_O4359  127 SSSSSEQRLL  136 + 10.6\n",
      "P22681_P0733  836 SDVWSYGILL  845 + 28.1\n",
      "P22681_Q8TBZ  763 SNLLHQMCLQ  772 + 7.85\n",
      "P22681_P2931  798 SDVWSFGIVM  807 + 26.3\n",
      "P22681_P1794 1080 SDVWSYGVLL 1089 + 28.5\n",
      "P22681_Q96JA  841 SYLSSQGTLS  850 + 8.45\n",
      "P22681_P4222  552 SVSWSQFNRE  561 + 8.00\n",
      "P22681_P1917  391 SDVL.HTIKE  399 + 8.11\n",
      "P22681_P1072  850 SDVWSYGIFL  859 + 25.5\n",
      "P22681_P0462  903 SDVWSYGVTV  912 + 22.5\n",
      "P22681_P0961  884 SDVWSFGILL  893 + 28.5\n",
      "P22681_P4023   21 AQVSSQDVSL   30 + 4.92\n",
      "P22681_Q0255  267 TRKL.FGHLE  275 + 3.88\n",
      "P22681_P4610   56 SRVSHYIINS   65 + 12.8\n",
      "P22681_P3178   47 TDSLSVSTLP   56 + 5.93\n",
      "P22681_Q9UGK  329 SSSW.PVILK  337 + 7.55\n",
      "P22681_P2932  819 SDVWSFGVVM  828 + 26.7\n",
      "P22681_P0612  301 CDSSSARSSL  310 + 6.47\n",
      "P22681_O9564  175 SSLSSRSCNS  184 + 8.21\n",
      "P22681_Q0276 1019 SDVWSYGVLL 1028 + 28.5\n",
      "P22681_P1231  112 SDPVHLTVLS  121 + 12.4\n",
      "P22681_Q7Z69  149 TDSSSNSSQK  158 + 7.31\n",
      "P22681_P0408  143 SRVHHRSVKV  152 + 3.88\n",
      "P22681_Q1415  144 ARPSSHRIKS  153 + 8.47\n",
      "P22681_P0794  453 SDVWSFGILQ  462 + 27.1\n",
      "P22681_P1356  908 TSTSSYYVFY  917 + 9.14\n",
      "P22681_Q9UM7 1310 TDTWSFGVLL 1319 + 24.0\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "10: Q99942\n",
      "\tMotif's length: 9\n",
      "\n",
      "                  *********\n",
      "Q99942_Q7Z43    7 KTYKYICRN   15 + 10.0\n",
      "Q99942_Q9P05   48 TNAQKYCQP   56 + 6.95\n",
      "Q99942_P4902  576 QNDKPYCQN  584 + 17.8\n",
      "Q99942_P1356 1177 KSTKPY.KN 1184 + 13.7\n",
      "Q99942_Q86WV  286 EQAKLFCRT  294 + 9.09\n",
      "Q99942_P6803  131 KDRKKFCKN  139 + 18.3\n",
      "Q99942_O4351  531 KSTKNY.KN  538 + 12.1\n",
      "\n",
      "\tMSA sequences length range: (8, 9)\n",
      "\n",
      "11: Q8IUQ4\n",
      "\tMotif's length: 7\n",
      "\n",
      "                  *******\n",
      "Q8IUQ4_Q9HB7   60 PAAVVAP   66 + 5.46\n",
      "Q8IUQ4_Q1555   20 PAASQPR   26 + 12.0\n",
      "Q8IUQ4_Q9Y6H  778 P.QQPSP  783 + 13.8\n",
      "Q8IUQ4_P3784  115 DMPVDPD  121 + 4.38\n",
      "Q8IUQ4_P3522   11 DMAMEPD   17 + 9.12\n",
      "Q8IUQ4_Q9H2X 1150 PRVLPSP 1156 + 15.1\n",
      "Q8IUQ4_Q9UPY  134 D.VAPPP  139 + 8.98\n",
      "Q8IUQ4_Q9GZT   80 PAAVPPP   86 + 14.9\n",
      "Q8IUQ4_P2959   35 PSPSPSP   41 + 13.8\n",
      "Q8IUQ4_P5182  589 PAQQEPP  595 + 15.8\n",
      "Q8IUQ4_Q1666  590 P.ESASP  595 + 10.2\n",
      "Q8IUQ4_Q9NWT   43 PRLSQSD   49 + 8.60\n",
      "Q8IUQ4_P4975  465 P.VLQPP  470 + 12.9\n",
      "Q8IUQ4_P9816 1148 PHPLPSP 1154 + 13.3\n",
      "Q8IUQ4_P0824  240 PEKQPAP  246 + 13.4\n",
      "Q8IUQ4_P1024  365 PEESASP  371 + 12.7\n",
      "Q8IUQ4_Q96RU  319 PMPLAPT  325 + 11.2\n",
      "Q8IUQ4_Q9P21 1141 PNPLPSR 1147 + 9.55\n",
      "Q8IUQ4_O7552  256 P.PPPPP  261 + 10.4\n",
      "Q8IUQ4_Q9970  229 D.QSQSP  234 + 9.90\n",
      "Q8IUQ4_Q1311  294 P.PSQPP  299 + 14.9\n",
      "Q8IUQ4_P4314 1348 P.LLPPP 1353 + 11.4\n",
      "Q8IUQ4_Q1663    9 PEQAPAP   15 + 12.9\n",
      "Q8IUQ4_Q1480  413 PEPMAAP  419 + 13.0\n",
      "Q8IUQ4_P4159 1016 P.RSPSP 1021 + 11.2\n",
      "Q8IUQ4_P4967  279 P.ASEAR  284 + 7.51\n",
      "Q8IUQ4_O0047  376 P.PLPST  381 + 9.38\n",
      "Q8IUQ4_P3688   54 PMVSESP   60 + 16.8\n",
      "Q8IUQ4_Q9993   30 PRQSEPP   36 + 16.1\n",
      "Q8IUQ4_Q96KS  399 P.VSQPP  404 + 14.0\n",
      "Q8IUQ4_P9817  374 PMVQEAI  380 + 10.1\n",
      "Q8IUQ4_Q6IE8  294 PEKMEPI  300 + 9.81\n",
      "Q8IUQ4_Q0791  725 PAPSPSP  731 + 17.6\n",
      "Q8IUQ4_P3727  851 PAVQEPP  857 + 15.6\n",
      "\n",
      "\tMSA sequences length range: (6, 7)\n",
      "\n",
      "12: Q969V5\n",
      "\tMotif's length: 9\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                  *****..****\n",
      "Q969V5_P0463  218 VPYEP..PEVG  226 + 22.0\n",
      "Q969V5_P3174  447 ITITP..PDQD  455 + 8.29\n",
      "Q969V5_O7538  500 RPYTPS.PQVG  509 + 14.7\n",
      "Q969V5_O1535  238 VPYEP..PQVG  246 + 23.5\n",
      "Q969V5_O0042  610 IPIMPASPQKG  620 + 12.2\n",
      "Q969V5_Q969V  208 VRLQP..PKQG  216 + 11.2\n",
      "\n",
      "\tMSA sequences length range: (9, 11)\n",
      "\n",
      "13: Q9NTX7\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  **********\n",
      "Q9NTX7_P1888  447 SPQKPPTPEE  456 + 16.6\n",
      "Q9NTX7_P1295  280 ALKPPPI.KL  288 + 5.72\n",
      "Q9NTX7_O1516   20 APRPPVPGEE   29 + 22.0\n",
      "Q9NTX7_Q9Y2T   21 APRPPVPGEE   30 + 22.0\n",
      "Q9NTX7_Q9H2G  286 SVQPHSTAEL  295 + 7.70\n",
      "Q9NTX7_O1523  641 ALPPPPPPHL  650 + 21.3\n",
      "Q9NTX7_O9527   26 APPPPPPPPL   35 + 23.1\n",
      "Q9NTX7_P0987  367 ATPPPSTASA  376 + 10.4\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "14: Q96PU5\n",
      "\tMotif's length: 7\n",
      "\n",
      "                  *****.....**\n",
      "Q96PU5_P6274   69 LRPLS.....YP   75 + 10.8\n",
      "Q96PU5_Q9NV9  171 YPVPPPY...SV  179 + 11.0\n",
      "Q96PU5_P3708  191 LRVPP.....PP  197 + 14.3\n",
      "Q96PU5_Q96PU  190 LPPPPL....PP  197 + 19.0\n",
      "Q96PU5_Q0195   49 NPRQS.....PV   55 + 10.6\n",
      "Q96PU5_P4928   33 NPSLSQ....SP   40 + 11.1\n",
      "Q96PU5_P5117  619 VPGTP.....PP  625 + 15.0\n",
      "Q96PU5_P0462  748 RPRAC.....PP  754 + 12.5\n",
      "Q96PU5_Q9BSA  504 SPPPTY....SP  511 + 10.5\n",
      "Q96PU5_Q9981  204 YPSQP.....PV  210 + 13.3\n",
      "Q96PU5_O0014   72 NPSPPP....SP   79 + 15.2\n",
      "Q96PU5_Q1452 1010 SPPPPETEKVPP 1021 + 14.6\n",
      "Q96PU5_Q96QD  241 FQVPC.....PV  247 + 7.64\n",
      "Q96PU5_Q1579  172 TPVLP.....PV  178 + 12.7\n",
      "Q96PU5_P5179  736 SPPLPPYT..PP  745 + 15.3\n",
      "Q96PU5_P3689  118 SPGLG.....PV  124 + 10.8\n",
      "Q96PU5_Q0791  626 ARPLP.....PP  632 + 14.7\n",
      "Q96PU5_Q8WXF   46 APPAPA....PP   53 + 15.8\n",
      "Q96PU5_O0021  287 SPSQGS....SP  294 + 10.1\n",
      "Q96PU5_P4300  232 VPVPG.....SV  238 + 10.9\n",
      "Q96PU5_P2200    8 LRSPP.....PP   14 + 15.2\n",
      "Q96PU5_Q6PJG  375 LQPLPQRP..SP  384 + 8.17\n",
      "Q96PU5_Q1280  960 SPRPPGE...PP  968 + 16.3\n",
      "Q96PU5_P3717  173 LPPLGVAI..SV  182 + 7.68\n",
      "Q96PU5_P5116  612 IPGTP.....PP  618 + 12.9\n",
      "Q96PU5_Q1662  774 RPRTC.....PQ  780 + 4.97\n",
      "Q96PU5_Q9H3D  538 TPPPP.....YP  544 + 17.0\n",
      "Q96PU5_Q1662  308 VPSPPSD...YV  316 + 12.5\n",
      "Q96PU5_O1496  295 MPSASSA...PP  303 + 10.8\n",
      "Q96PU5_Q9H1D   18 APRLS.....PV   24 + 11.9\n",
      "Q96PU5_Q4VCS  837 VPSTP.....SP  843 + 14.6\n",
      "Q96PU5_P5380  163 SPPAS.....PP  169 + 15.5\n",
      "Q96PU5_Q96D3    6 APPPSRS...SP   14 + 13.5\n",
      "Q96PU5_O1464  685 MPPPP.....PP  691 + 20.1\n",
      "Q96PU5_Q6UUV  443 SPPPP.....YP  449 + 18.9\n",
      "Q96PU5_O7547  885 VPSPA.....YP  891 + 10.9\n",
      "Q96PU5_Q9BXB   25 APPLCA....AP   32 + 8.67\n",
      "Q96PU5_Q9P2K  664 GPGTP.....PP  670 + 11.1\n",
      "\n",
      "\tMSA sequences length range: (7, 12)\n",
      "\n",
      "15: Q96J02\n",
      "\tMotif's length: 8\n",
      "\n",
      "                  ********\n",
      "Q96J02_P0815 1074 TPPPSGPP 1081 + 15.0\n",
      "Q96J02_O9583  244 SVTPPPPP  251 + 20.0\n",
      "Q96J02_Q1357  252 GGPAPPPY  259 + 16.4\n",
      "Q96J02_O1464  686 PPPPPPVP  693 + 20.2\n",
      "Q96J02_Q0186   61 PPCPPPYE   68 + 11.0\n",
      "Q96J02_Q1662   77 PLPPPPYE   84 + 16.0\n",
      "Q96J02_Q9Y5X  147 AFGHPQAY  154 + 8.17\n",
      "Q96J02_Q1579  218 PETPPPGY  225 + 15.7\n",
      "Q96J02_P4598   20 TPGPVGSP   27 + 11.0\n",
      "Q96J02_Q9H3M  327 TPEAPPCY  334 + 9.91\n",
      "Q96J02_Q1663  315 GPPPPPGP  322 + 19.6\n",
      "Q96J02_O9599  198 PGAPPLPP  205 + 15.8\n",
      "Q96J02_P5595   42 HELPVLAP   49 + 8.83\n",
      "Q96J02_Q1530 1294 TVLPPPPY 1301 + 16.7\n",
      "Q96J02_Q1451  298 SPNHP.PP  304 + 10.5\n",
      "Q96J02_Q96J0  257 SRPPPPTP  264 + 13.9\n",
      "Q96J02_P4653 2407 SLQPPPPP 2414 + 20.2\n",
      "Q96J02_Q9H3D  536 HCTPPPPY  543 + 18.9\n",
      "Q96J02_Q1311  471 ALPPTPAP  478 + 15.6\n",
      "Q96J02_Q1665  109 AEQSPPPY  116 + 11.7\n",
      "Q96J02_O1551  216 QLGAQQEP  223 + 4.84\n",
      "Q96J02_P1727  175 GPEPPPVY  182 + 16.3\n",
      "Q96J02_P0541  237 GETPPLSP  244 + 12.8\n",
      "Q96J02_Q9HBA  137 APAPQPPP  144 + 17.5\n",
      "Q96J02_P2268  541 DLPPPPPP  548 + 19.8\n",
      "Q96J02_O1535 1094 HPRPPLPP 1101 + 15.8\n",
      "Q96J02_Q1558   78 SCAPPPPP   85 + 19.0\n",
      "Q96J02_O4335    8 SALPTIPY   15 + 8.05\n",
      "Q96J02_Q9Y37  267 SVTPVPSV  274 + 8.25\n",
      "Q96J02_Q9UM4   16 SPPPPPPP   23 + 23.7\n",
      "Q96J02_Q1354  362 SLEHPQEE  369 + 6.94\n",
      "Q96J02_P1917  144 TLQAP.TP  150 + 5.86\n",
      "Q96J02_P0624   33 DPTPQHYP   40 + 5.64\n",
      "Q96J02_O1496  724 SLPPQQPY  731 + 15.6\n",
      "Q96J02_Q7Z43  117 PGPPTPAA  124 + 10.7\n",
      "Q96J02_O4331  117 HGAEPLPY  124 + 6.65\n",
      "Q96J02_Q86Y0  231 PLPPPPPP  238 + 22.5\n",
      "Q96J02_Q6PJG  331 TVQAPGPP  338 + 12.5\n",
      "Q96J02_P1784  560 TGNPTGTY  567 + 7.82\n",
      "Q96J02_Q1662    5 PLESPPPY   12 + 14.8\n",
      "Q96J02_Q9996  276 GLSHTGTP  283 + 8.01\n",
      "Q96J02_Q9UBN  396 GPPPTIVE  403 + 7.78\n",
      "Q96J02_O6024  461 SPGAPPGP  468 + 15.2\n",
      "Q96J02_Q9954    5 AVDPQPSV   12 + 5.66\n",
      "Q96J02_Q1349  134 SFAHSLSP  141 + 5.53\n",
      "Q96J02_Q8NEA  300 PPGPPPPY  307 + 20.9\n",
      "Q96J02_Q9BTU   45 GPSPPGSP   52 + 13.9\n",
      "Q96J02_O1510  203 ELESPPPP  210 + 11.5\n",
      "Q96J02_Q1363 1382 HPPPVPGP 1389 + 17.0\n",
      "Q96J02_Q1277  119 HPAPPQPP  126 + 19.3\n",
      "Q96J02_P1640    9 PAAAPPAE   16 + 10.0\n",
      "Q96J02_P1505  397 GLSATPPA  404 + 7.45\n",
      "Q96J02_Q1575  374 QPTPSPAP  381 + 13.9\n",
      "Q96J02_Q8WWW   31 PELPPPPP   38 + 18.0\n",
      "Q96J02_Q9UMX   11 GPTAPPAP   18 + 16.7\n",
      "Q96J02_P2096   58 SADAP.AY   64 + 2.71\n",
      "Q96J02_P2186 1207 SPPHPPRP 1214 + 13.0\n",
      "Q96J02_O7558 1566 PVPPPPTP 1573 + 19.2\n",
      "Q96J02_Q1581  294 PLPPVLPP  301 + 15.3\n",
      "Q96J02_P4693    6 QPPPQPAP   13 + 17.1\n",
      "Q96J02_Q4VCS 1013 TPAPTPTP 1020 + 16.2\n",
      "\n",
      "\tMSA sequences length range: (7, 8)\n",
      "\n",
      "16: O00308\n",
      "\tMotif's length: 8\n",
      "\n",
      "                  *****.........................***\n",
      "O00308_P4928  339 SPHAGLF.......................PKD  348 + 6.24\n",
      "O00308_O1510  206 SPPPPYSRY.....................PMD  217 + 19.5\n",
      "O00308_P8402  179 TPPPGYLSEDGETSDHQMNHSMDAGSPNLSPNP  211 + 14.6\n",
      "O00308_Q9H07   65 TPPPGAQCEVP...................ASP   78 + 14.7\n",
      "O00308_Q9UM4   16 SPPPP.........................PPP   23 + 24.4\n",
      "O00308_Q0186   12 SPPPGGGGDGPGG.................PEP   27 + 19.1\n",
      "O00308_P2492 1566 SPTPGSPGS.....................PGP 1577 + 13.0\n",
      "O00308_O0030  255 SPPAAPLSVT....................PNP  267 + 13.4\n",
      "O00308_P1116  166 SPPPP.........................PPP  173 + 24.4\n",
      "O00308_Q9P2A  263 SPPPPDEELPLPLDL...............PPP  280 + 21.9\n",
      "O00308_O1535  167 TPPPPGTAIR....................AMP  179 + 17.8\n",
      "O00308_Q8IUC  359 SPPPP.........................PPP  366 + 24.4\n",
      "\n",
      "\tMSA sequences length range: (8, 33)\n",
      "\n",
      "17: O43791\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  ***...................*******\n",
      "O43791_Q9Y6Q 1282 PPPNVTASPSMDGLLAGPTMPQAPPQQFP 1310 + 13.1\n",
      "O43791_P3565  192 PLP...................KSKKTCS  201 + 6.74\n",
      "O43791_Q1280  959 SSP...................RPPGEPP  968 + 16.0\n",
      "O43791_Q9UBN  802 PPPLLT................LPRPPLS  814 + 18.5\n",
      "O43791_Q9UER  701 PSPARLSQTP............HSQPPRP  717 + 17.4\n",
      "O43791_P0337  105 PSPLMLL...............HPPPQLS  118 + 19.7\n",
      "O43791_P3522  273 PSPSTPVQ..............SPHPQFP  287 + 19.6\n",
      "O43791_P0541  206 QPPH..................HLPQQMP  216 + 14.8\n",
      "O43791_O7536  130 PPPAKKAK..............SPSQKKP  144 + 19.3\n",
      "O43791_P0815  868 PPPDY.................LPSEPRP  879 + 18.1\n",
      "O43791_P1027  374 PPP...................PPPPPHP  383 + 23.0\n",
      "O43791_P6048  246 PLPVCGDIKVEFF.........HKQNKML  265 + 2.11\n",
      "O43791_Q9HCU  196 SLPP..................SKRKKAP  206 + 9.27\n",
      "O43791_O0033  150 PPPVS.................KPSEPRP  161 + 18.1\n",
      "O43791_P1007 1313 PPP...................HPVQSYP 1322 + 13.5\n",
      "O43791_P1007  377 SPPLI.................HPAPTFP  388 + 19.0\n",
      "O43791_Q9BYW  185 SPPS..................SPPPPPP  195 + 23.0\n",
      "O43791_Q27J8  438 PPP...................PPPPPLP  447 + 24.6\n",
      "O43791_Q9H9S   53 PLPSSMDLLIQDSPDSST....SPKGKQP   77 + 11.5\n",
      "O43791_Q1283   40 PSPMRAANRS............HSAGRTP   56 + 11.2\n",
      "O43791_P2544   62 PPPE..................VSNPKKP   72 + 12.6\n",
      "O43791_O6088  759 PPPQQ.................PPPPPPP  770 + 25.0\n",
      "O43791_O1516  512 PPPRLINFQNH...........SPKPNGP  529 + 12.9\n",
      "O43791_P2486  388 PLPSG.................LLTPPQS  399 + 7.84\n",
      "O43791_Q9H5J  227 PSPTV.................RPPDEPP  238 + 16.4\n",
      "O43791_Q1290  338 PLP...................SKMPKKP  347 + 13.5\n",
      "O43791_P4932  410 PPPAP.................APHATLP  421 + 15.6\n",
      "O43791_Q1682  194 SSP...................SSSPPTS  203 + 12.6\n",
      "O43791_Q49A2  166 QSP...................RKRGRPP  175 + 10.6\n",
      "O43791_P0640  422 PPPL..................PPRATPS  432 + 16.2\n",
      "\n",
      "\tMSA sequences length range: (10, 29)\n",
      "\n",
      "18: Q53G59\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  *****..........*****\n",
      "Q53G59_P2191  294 CAPPA..........PGLPP  303 + 20.1\n",
      "Q53G59_P1463   28 PTAPAATSK......PGLRP   41 + 13.6\n",
      "Q53G59_O9497  944 PPPPSSGASFQHGG.PGAPP  962 + 23.2\n",
      "Q53G59_Q9NQW  955 PAPPASFPVPYLPGDPGAPC  974 + 18.1\n",
      "Q53G59_Q9299  670 PPPPAAMGP......PGAPP  683 + 27.9\n",
      "Q53G59_Q9UBV   11 PGAAGQA........PGAPP   22 + 19.0\n",
      "Q53G59_Q9C0E  195 PGPPKDSSA......PGGPP  208 + 18.2\n",
      "Q53G59_Q9294  398 PGPPGG.........PGMPP  408 + 24.6\n",
      "\n",
      "\tMSA sequences length range: (10, 20)\n",
      "\n",
      "19: Q7L5Y6\n",
      "\tMotif's length: 10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                  **********\n",
      "Q7L5Y6_Q1308 1998 PADPANLDSE 2007 + 21.5\n",
      "Q7L5Y6_P0463   87 PAPSWPLSSS   96 + 12.1\n",
      "Q7L5Y6_Q1277  125 PPPPGPLSQH  134 + 17.2\n",
      "Q7L5Y6_P0541  240 PPLS.PIDME  248 + 8.81\n",
      "Q7L5Y6_O0076 2109 PADPANLDSE 2118 + 21.5\n",
      "Q7L5Y6_Q1333  700 PPPPAPVNDE  709 + 20.6\n",
      "Q7L5Y6_Q53ET  285 PPLPTPLDPE  294 + 19.9\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "20: Q9Y4B6\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  **********\n",
      "Q9Y4B6_O7544  322 EHEASR.RVK  330 + 10.1\n",
      "Q9Y4B6_P1305   67 SAEQLD.RIQ   75 + 11.2\n",
      "Q9Y4B6_Q7L59  160 LKERRVQRIQ  169 + 16.8\n",
      "Q9Y4B6_Q9Y3Z  610 LREASKSRVQ  619 + 16.8\n",
      "Q9Y4B6_P3015  411 LAEDAKWRVR  420 + 9.23\n",
      "Q9Y4B6_P3524  404 EAEQEMQRIK  413 + 15.5\n",
      "Q9Y4B6_O1474  650 RAERLTSRVK  659 + 15.7\n",
      "Q9Y4B6_O9583  623 DEERRESRIQ  632 + 21.6\n",
      "Q9Y4B6_Q9NRM  586 DEEKRESRIK  595 + 20.0\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "21: Q96SW2\n",
      "\tMotif's length: 7\n",
      "\n",
      "                  *******\n",
      "Q96SW2_Q96SW  389 AQCKICA  395 + 16.0\n",
      "Q96SW2_Q9H3D  340 FEARICA  346 + 13.9\n",
      "Q96SW2_Q9UJQ  594 FQCKICG  600 + 21.6\n",
      "Q96SW2_P1063  604 VQSK.CG  609 + 5.51\n",
      "Q96SW2_Q1342  490 FECNMCG  496 + 14.6\n",
      "\n",
      "\tMSA sequences length range: (6, 7)\n",
      "\n",
      "22: Q969P5\n",
      "\tMotif's length: 10\n",
      "\n",
      "                  **********\n",
      "Q969P5_Q1655  137 SAP.RGNETS  145 + 2.00\n",
      "Q969P5_O0030   27 PAPTPAPAAA   36 + 18.4\n",
      "Q969P5_P1517  214 SGP.PSGARR  222 + 12.3\n",
      "Q969P5_Q1277   40 SSPAPSGSAA   49 + 14.7\n",
      "Q969P5_P1517  154 PQP.GVPSEC  162 + 12.6\n",
      "Q969P5_P2856  318 SAEAGSPAMA  327 + 9.16\n",
      "Q969P5_P1942   12 PRPAPAPIRR   21 + 12.9\n",
      "Q969P5_Q9UBF   21 LQP.PGPETS   29 + 14.4\n",
      "Q969P5_Q9Y6J 1818 PAPAPAPATT 1827 + 22.7\n",
      "Q969P5_Q1489  104 LAPAPAPAEA  113 + 22.9\n",
      "Q969P5_P0110   40 LQP.PAPSED   48 + 14.2\n",
      "Q969P5_O4347  139 SGPASAPSTC  148 + 16.9\n",
      "\n",
      "\tMSA sequences length range: (9, 10)\n",
      "\n",
      "23: P19474\n",
      "\tMotif's length: 9\n",
      "\n",
      "                 *********\n",
      "P19474_Q0255 222 TCPEGCRLS 230 + 13.8\n",
      "P19474_Q1310 460 VCPECAKVS 468 + 11.6\n",
      "P19474_O1492 697 SLPEPAKKS 705 + 15.5\n",
      "P19474_P4652  98 ACKVPAQES 106 + 11.4\n",
      "P19474_Q9C03  54 SCPV.CRIS  61 + 12.9\n",
      "P19474_P1947  50 VCPV.CRQR  57 + 12.4\n",
      "P19474_Q1465 162 VAPEPCPQP 170 + 19.3\n",
      "P19474_Q9H21  90 ACPSPGQKI  98 + 11.8\n",
      "P19474_Q9298 242 TTPSPGPQP 250 + 13.5\n",
      "P19474_Q0420 340 SVPKPAPQP 348 + 16.3\n",
      "\n",
      "\tMSA sequences length range: (8, 9)\n"
     ]
    }
   ],
   "source": [
    "urlbase = \"https://awi.cuhk.edu.cn/~ubinet\"\n",
    "counter = 0       # monitor\n",
    "\n",
    "# Visualize MSAs from the motifs HTML of UbiNet directly\n",
    "with open(url_motifs_path) as file:\n",
    "    for url in file:\n",
    "        if url.split(\"/\")[3] in lowq_E3s:\n",
    "            \n",
    "            counter += 1\n",
    "            print(f'\\n{counter}: {url.split(\"/\")[3]}')\n",
    "            \n",
    "            # Load corresponding PWM\n",
    "            weight_m = pd.read_csv(weight_m_path+url.split(\"/\")[3]+\".tsv\", sep = \"\\t\")\n",
    "            print(f\"\\tMotif's length: {len(weight_m.index)}\")\n",
    "            \n",
    "            # Download corresponding MSA\n",
    "            r = (requests.get(urlbase+url.strip())).text              \n",
    "            soup = BeautifulSoup(r, 'html.parser')\n",
    "            msa = str(soup.find_all(name = 'input', attrs = {\"name\": \"aln1\"}))\n",
    "            print(msa.split('value=\".')[1].split('.\"/>')[0])\n",
    "            \n",
    "            # Print MSA's sequences length range\n",
    "            print(f'\\tMSA sequences length range: {lowq_E3s[url.split(\"/\")[3]]}')\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87a1ef4d-111c-48bc-8c93-5636363839fa",
   "metadata": {},
   "source": [
    "## 3. MSAs filtering (refinement)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70bc0eeb-878b-45ec-b9b9-f0c249f9268d",
   "metadata": {},
   "source": [
    "After previous MSA analysis and visualization, we decided to filter the positive sequences to preserve only those contributing completely to motif generation, which are those of the same length as the motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "589b7d3c-74aa-4b23-b486-2c9234a6ffca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maintain only MSA sequences of the same length as the motif\n",
    "\n",
    "filter_motifs_msa(ubinet_degrons_path, weight_m_path, ubinet_degrons_kept_path, ubinet_degrons_discarded_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
